File: ValidateBinarySearchTree2.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        
        vector<TreeNode*> stack;
        TreeNode *p=root;
        TreeNode *pre=NULL;
        while(p!=NULL||!stack.empty()){
            while(p!=NULL){
                stack.push_back(p);
                p=p->left;
            }
            p=stack.back();
            stack.pop_back();
            if(pre!=NULL&&p->val<=pre->val) return false;
            pre=p;
            p=p->right;
        }
        
        return true;
    }
};

int main(){
    TreeNode *root=new TreeNode(10);
    TreeNode *a=new TreeNode(5);
    TreeNode *b=new TreeNode(15);
    TreeNode *c=new TreeNode(6);
    TreeNode *d=new TreeNode(20);
    TreeNode *e=NULL;
    root->left=a;
    root->right=b;
    b->left=c;
    b->right=d;
    
    Solution *solution=new Solution();
    solution->isValidBST(root) ? cout<<"True" : cout<<"False";
    
    return 0;
    
}

File: ValidateBinarySearchTree2.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def is_validBST(root, minNode, maxNode):
            if root==None:
                return True
            if (minNode!=None and root.val<=minNode.val) or (maxNode!=None and root.val>=maxNode.val):
                return False
            return is_validBST(root.left, minNode, root) and is_validBST(root.right, root, maxNode)
        return is_validBST(root, None, None)
        
if __name__=='__main__':
    root = TreeNode(10);
    a = TreeNode(5);
    b = TreeNode(15);
    c = TreeNode(6);
    d = TreeNode(20);
    root.left = a
    root.right = b
    b.left = c
    b.right = d
    
    solution = Solution()
    print(solution.isValidBST(root))
